File: Scala/0002.scala
object Solution {
    def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
        val dummyHead = new ListNode(0)
        var current = dummyHead
        var carry = 0

        var p = l1
        var q = l2

        while (p != null || q != null || carry != 0) {
            val x = if (p != null) p.x else 0
            val y = if (q != null) q.x else 0
            
            val sum = x + y + carry
            carry = sum / 10
            
            current.next = new ListNode(sum % 10)
            current = current.next
            
            if (p != null) p = p.next
            if (q != null) q = q.next
        }

        dummyHead.next
    }
}

File: Java/0002.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        // Create a dummy node to simplify list operations
        ListNode dummyHead = new ListNode(0);
        ListNode current = dummyHead;
        int carry = 0;

        // Traverse both linked lists
        while (l1 != null || l2 != null || carry != 0) {
            // Get the current values from both nodes
            int val1 = (l1 != null) ? l1.val : 0;
            int val2 = (l2 != null) ? l2.val : 0;

            // Calculate the sum of current digits and carry
            int sum = val1 + val2 + carry;
            carry = sum / 10;
            sum = sum % 10;

            // Create a new node with the sum value
            current.next = new ListNode(sum);
            current = current.next;

            // Move to the next nodes in the lists
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }

        // Return the next of dummy node, which is the head of the result list
        return dummyHead.next;
    }
}